### PR TITLE
Implement JWT token type for Applications

### PR DESCRIPTION
**Pull request by tharindu1st**
## Purpose
Implement JWT token type for Applications

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/auth/dto/DCRClientInfo.java [hunk 1]
```diff
@@ -161,6 +164,16 @@ public void setUserinfoSignedResponseAlg(String userinfoSignedResponseAlg) {
...
     }*/
 
+    public String getTokenType() {
+
+        return tokenType;
+    }
+
+    public void setTokenType(String tokenType) {
+
+        this.tokenType = tokenType;
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/auth/dto/DCRClientInfo.java]: please remove above extra lines.

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java [hunk 2]
```diff
@@ -496,17 +498,28 @@ public OAuthApplicationInfo updateGrantTypesAndCallbackURL(String applicationId,
...
         }
 
-        if (grantTypes == null || grantTypes.isEmpty() || StringUtils.isEmpty(callbackURL)) {
+        if (grantTypes == null || grantTypes.isEmpty()) {
             String msg = "Both Grant Types list and Callback URL can't be null or empty at once.";
             log.error(msg);
             throw new APIManagementException(msg, ExceptionCodes.OAUTH2_APP_RETRIEVAL_FAILED);
         }
+        if (grantTypes.contains(APIMgtConstants.Oauth2Constants.AUTHORIZATION_CODE_GRANT_TYPE) || grantTypes.contains
+                (APIMgtConstants.Oauth2Constants.CODE_GRANT_TYPE)) {
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]: what's the difference between these 2?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java [hunk 3]
```diff
@@ -496,17 +498,28 @@ public OAuthApplicationInfo updateGrantTypesAndCallbackURL(String applicationId,
...
+            if (StringUtils.isEmpty(callbackURL)) {
+                String msg = "Callback URL can't be null or empty when" + APIMgtConstants.Oauth2Constants
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]:
```suggestion
                String msg = "Callback URL can't be null or empty when " + APIMgtConstants.Oauth2Constants
```

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java [hunk 4]
```diff
@@ -496,17 +498,28 @@ public OAuthApplicationInfo updateGrantTypesAndCallbackURL(String applicationId,
...
+                        .AUTHORIZATION_CODE_GRANT_TYPE + " or" + APIMgtConstants.Oauth2Constants.CODE_GRANT_TYPE + " " +
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]:
```suggestion
                        .AUTHORIZATION_CODE_GRANT_TYPE + " or " + APIMgtConstants.Oauth2Constants.CODE_GRANT_TYPE + " " +
```

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java (hunk 2 above)
Comment (Rajith90) [reply components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]: This should be implicit and authorization code grants .

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorService.java [hunk 5]
```diff
@@ -473,8 +473,8 @@ private OAuthApplicationInfo createDCRApplication(String clientName, String call
             // Here the keyType:"Application" will be passed as a default value
             // for the oAuthAppRequest constructor argument.
             // This value is not related to DCR application creation.
-            OAuthAppRequest oAuthAppRequest = new OAuthAppRequest(clientName,
-                    callBackURL, AuthenticatorConstants.APPLICATION_KEY_TYPE, grantTypes);
+            OAuthAppRequest oAuthAppRequest = new OAuthAppRequest(clientName, callBackURL, AuthenticatorConstants
+                    .APPLICATION_KEY_TYPE, grantTypes, "OAUTH");
```
[the remark this answers is not in the document]
Comment (tharindu1st) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.authenticator/src/main/java/org/wso2/carbon/apimgt/rest/api/authenticator/AuthenticatorService.java]: TokenType enum is in a different package

**Pull request merged**